lib/roo_on_rails/concerns/require_api_key.rb

```diff
@@ -0,0 +1,108 @@
+require 'active_support/concern'
+require 'action_controller/metal/http_authentication'
+
+module RooOnRails
+  module Concerns
+    # This concern allows API authentication in a consistent manner.
+    #
+    # If a service connects with basic auth using the username "service" then the
+    # `SERVICE_CLIENT_KEY` environment variable must have the given password as one of
+    # the comma separated strings within it or a 403 will be raised.
+    #
+    # @example: Any service with an acceptable key can access routes
+    #
+    #     class ThingController < ApplicationController
+    #       include RooOnRails::Concerns::RequireApiKey
+    #       require_api_key
+    #
+    #       # etc
+    #     end
+    #
+    # @example: Only the specified clients can access specific routes in this controller
+    #
+    #     class ThingController < ApplicationController
+    #       include RooOnRails::Concerns::RequireApiKey
+    #       require_api_key(only_services: :my_service, only: :create)
+    #
+    #       # etc
+    #     end
+    module RequireApiKey
+      extend ActiveSupport::Concern
+      include ActionController::HttpAuthentication::Basic::ControllerMethods
+
+      attr_reader :current_client
+
+      module ClassMethods
+        # Declares that routes on the controller must have access credentials specified
+        # in the request that match the approparite environment variables.
+        #
+        # @param :only_services (#to_s,Array<#to_s>) Restricts the services which will be accepted
+        # @see AbstractController::Callbacks::ClassMethods#before_action for additional scoping opts
+        def require_api_key(only_services: nil, **options)
+          before_action(**options) do
+            authenticate_or_request_with_http_basic('actor-tracking') do |service_name, client_key|
+              Authenticator.new(*only_services).valid?(service_name, client_key).tap do |is_valid|
```

Comment:
If `only_services` is nil will this cause an exception?

Reply:
`*nil` -> `[]`

Reply:
Good spot!